include/fastad_bits/mat.hpp

```diff
@@ -0,0 +1,96 @@
+#pragma once
+#include <iostream>
+#include <vector>
+
+namespace ad {
+
+template <typename T>
+class Mat;
+template <typename T>
+std::ostream& operator<<(std::ostream& os, const Mat<T>& mat);
+template <typename T>
+bool operator==(const Mat<T>& mat1, const Mat<T>& mat2);
+
+template <typename T>
+class Mat {
+
+public:
+
+	Mat() : data(), rows_(0), cols_(0) {}
+	Mat(const Mat& mat) : data(mat.data), rows_(mat.rows_), cols_(mat.cols_) {}
+	Mat(size_t size, T fill) : data(size*size, fill), rows_(size), cols_(size) {}
+	Mat(size_t rows, size_t cols, T fill) : data(rows*cols, fill), rows_(rows), cols_(cols) {}
+
+	friend std::ostream& operator<<<>(std::ostream& os, const Mat& mat);
+	friend bool operator==<>(const Mat& mat1, const Mat& mat2);
+	friend bool operator!=(const Mat& mat1, const Mat& mat2) { return !(mat1 == mat2); }
+	T& operator()(size_t row, size_t col) { return data[row*cols_ + col];  }
+	const T& operator()(size_t row, size_t col) const { return ((Mat&)*this)(row, col); }
+
+	size_t size() const { return rows_*cols_; }
+	size_t n_rows() const { return rows_; }
+	size_t n_cols() const { return cols_; }
+
+	void print(std::ostream& os, std::string header) { os << header << std::endl << *this; }
+	void print(std::string header) { print(std::cout, header);  }
+
+	void fill(size_t rows, size_t cols, T fill) { *this = Mat(rows, cols, fill); }
```

Comment:
This unnecessarily invokes copy assignment. We can just directly modify `this->data`. Do `#include <algorithm>` at the top and write instead
```cpp
this->data.resize(rows * cols);
std::fill(this->data.begin(), this->data.end(), fill);
```